src/hotspot/share/opto/regmask.hpp

```diff
@@ -65,27 +68,266 @@ class RegMask {
+  // constructors. If we get copied/cloned, &_RM_UP_EXT will no longer equal
+  // orig_ext_adr.
+  uintptr_t** orig_ext_adr = &_RM_UP_EXT;
+  //
+  // If the original version, of which we may be a clone, is read-only. In such
+  // cases, we can allow read-only sharing.
+  bool orig_const = false;
+#endif
+
+  // Current total register mask size in words
+  unsigned int _rm_size;
+
+  // We support offsetting register masks to present different views of the
+  // register space, mainly for use in PhaseChaitin::Select. The _offset
+  // variable indicates how many words we offset with. We consider all
+  // registers before the offset to not be included in the register mask.
+  unsigned int _offset;
+
+  // If _all_stack = true, we consider all registers beyond what the register
+  // mask can currently represent to be included. If _all_stack = false, we
+  // consider the registers not included.
+  bool _all_stack = false;
+
+  // The low and high watermarks represent the lowest and highest word that
+  // might contain set register mask bits, respectively. We guarantee that
+  // there are no bits in words outside this range, but any word at and between
+  // the two marks can still be 0. We do not guarantee that the watermarks are
+  // optimal. If _hwm < _lwm, the register mask is necessarily empty. Indeed,
+  // when we construct empty register masks, we set _hwm = 0 and _lwm = max.
   unsigned int _lwm;
   unsigned int _hwm;
 
- public:
-  enum { CHUNK_SIZE = _RM_SIZE * BitsPerWord };
+  // The following diagram illustrates the internal representation of a RegMask
+  // (with _offset = 0, for a made-up platform with 10 registers and 4-bit
+  // words) that has been extended with two additional words to represent more
+  // stack locations:
+  //                                   _hwm=3
+  //            _lwm=1                RM_SIZE=3                _rm_size=5
+  //              |                       |                        |
+  //   r0 r1 r2 r3 r4 r5 r6 r7 r8 r9 s0 s1   s2 s3 s4 s5 s6 s7 s8 s9 s10 s11 ...
+  //  [0  0  0  0 |0  1  1  0 |0  0  1  0 ] [1  1  0  1 |0  0  0  0] as  as  as
+  // [0]         [1]         [2]           [0]         [1]
+  //
+  // \____________________________________/ \______________________/
+  //                        |                           |
+  //                      RM_UP                     RM_UP_EXT
+  // \_____________________________________________________________/
+  //                                 |
+  //                             _rm_size
+  //
+  // In this example, registers {r5, r6} and stack locations {s0, s2, s3, s5}
+  // are included in the register mask. Depending on the value of _all_stack,
+  // (s10, s11, ...) are all included (as = 1) or excluded (as = 0). Note that
+  // all registers/stack locations under _lwm and over _hwm are excluded.
+  // The exception is (s10, s11, ...), where the value is decided solely by
+  // _all_stack, regardless of the value of _hwm.
```

Comment:
This comment illustrates the case with `_offset = 0`, I think it would be useful to extend it with an example where `_offset > 0`. Here is a suggestion: https://github.com/openjdk/jdk/commit/8377012ac485a70703921822d58bc535bafb7a0a. Feel free to merge as-is or edit to your liking, if you agree.

Reply:
Looks good to me, now merged. There are likely other opportunities for more source code comment illustrations throughout `regmask.hpp`. `SUBTRACT_inner` and `overlap` comes to mind, in particular. I'll have a look and see what can be improved.

Reply:
Thanks, Daniel!